src/Common/GuardedPoolAllocatorCrashHandler.cpp

```diff
@@ -0,0 +1,285 @@
+    {
+        uintptr_t address = getAllocationAddress(allocation_meta);
+        size_t size = getAllocationSize(allocation_meta);
+        if (error == Error::USE_AFTER_FREE)
+        {
+            snprintf(
+                description_buffer,
+                kDescriptionBufferLen,
+                "(%zu byte%s into a %zu-byte allocation at 0x%zx)",
+                access_ptr - address,
+                (access_ptr - address == 1) ? "" : "s",
+                size,
+                address);
+        }
+        else if (access_ptr < address)
+        {
+            snprintf(
+                description_buffer,
+                kDescriptionBufferLen,
+                "(%zu byte%s to the left of a %zu-byte allocation at 0x%zx)",
+                address - access_ptr,
+                (address - access_ptr == 1) ? "" : "s",
+                size,
+                address);
+        }
+        else if (access_ptr > address)
+        {
+            snprintf(
+                description_buffer,
+                kDescriptionBufferLen,
+                "(%zu byte%s to the right of a %zu-byte allocation at 0x%zx)",
+                access_ptr - address,
+                (access_ptr - address == 1) ? "" : "s",
+                size,
+                address);
+        }
+        else
+        {
+            /// DOUBLE_FREE
+            snprintf(description_buffer, kDescriptionBufferLen, "(a %zu-byte allocation) ", size);
+        }
+    }
+
+    /// Possible number of digits of a 64-bit number: ceil(log10(2^64)) == 20. Add
+    /// a null terminator, and round to the nearest 8-byte boundary.
+    uint64_t thread_id = access_thread_id;
+    char thread_buffer[kThreadBufferLen];
+    if (thread_id == clickhouse_gwp_asan::kInvalidThreadID)
+        snprintf(thread_buffer, kThreadBufferLen, "<unknown>");
+    else
+        snprintf(thread_buffer, kThreadBufferLen, "%" PRIu64, thread_id);
+
+    LOG_FATAL(
+        log,
+        "{} at {} {} by thread {} here:\n",
+        clickhouse_gwp_asan::errorToString(error),
+        reinterpret_cast<void *>(access_ptr),
+        description_buffer,
+        thread_buffer);
+}
+
+void CrashReporter::dumpReport(uintptr_t error_ptr, uint32_t access_thread_id)
+{
+    assert(allocator_state && "missing allocator_state in report.");
+    assert(allocator_metadata && "missing allocator_metadata in report.");
+
+    if (!errorIsMine(error_ptr))
+        return;
+
+    ScopedEndOfReportDecorator decorator(log);
+
+    uintptr_t internal_error_ptr = getInternalCrashAddress();
+    if (internal_error_ptr != 0u)
+        error_ptr = internal_error_ptr;
+
+    Error error = diagnoseError(error_ptr);
+
+    if (error == Error::UNKNOWN)
+    {
+        LOG_FATAL(
+            log,
+            "GWP-ASan cannot provide any more information about this error. "
+            "This may occur due to a wild memory access into the GWP-ASan pool, "
+            "or an overflow/underflow that is > 512B in length.");
```

Comment:
I guess the correct value here should be `PAGE_SIZE / 2`.
The reasoning for this is the following: imagine we have an allocation, which resides at the end of allocated page (in `allocate` we randomly place allocation either in the beginning or the end of allocated chunk), let's say it has `slot_index = n`. Then imagine this allocation overflows by `> PAGE_SIZE / 2` and lands on an address, which belongs to a guard page. This causes a `SIGSEGV` signal with that addres.

After that we receive this address in crash handler and wrongfully attribute it to allocation in `slot_index = n + 1` in [addrToMetadata](https://github.com/ClickHouse/ClickHouse/pull/36826/files#diff-f8a56281e19db7a9019dc5014e63cf8143ed22c989236bff04788f632487d2b4R97) which calls [getNearestSlot](https://github.com/ClickHouse/ClickHouse/pull/36826/files#diff-f8a56281e19db7a9019dc5014e63cf8143ed22c989236bff04788f632487d2b4R99).
And so this causes us to look at metadata at `slot_index = n + 1`, which may contain metadata for another allocation or even be completely empty.

So we cannot guarantee anything of overflow / underflow `> PAGE_SIZE / 2`

> It should be possible even if it less, in case of allocation is too small (i.e. 4096-allocation_size)

This is true, for anything that is `< PAGE_SIZE / 2` we can diagnose it.
But whether allocation was placed at the beginning / end of allocated chunk should also be taken into consideration.

So if you agree with this reasoning I will correct it to `PAGE_SIZE / 2`

Reply:
Oh, I see, this is because of this code - https://github.com/ClickHouse/ClickHouse/pull/36826/files#diff-50d450453a74dec6b9b8a9634c7f60525f53d0f8fc9a85d61fcc16036bba400dR63

But overflow/underflow > PAGE_SIZE/2 is quite infrequently.

Anyway I see that this code initially came from llvm, so let's keep it as-is them for now.

Reply:
Ok